File: test-2025.9.2/编程/数独/数独.cpp
#include <iostream>
#include <vector>
#include <cmath>

bool is_valid(std::vector<std::vector<int>>& board, int row, int col, int num, int size) {
    for (int i = 0; i < size; i++) {
        if (board[row][i] == num || board[i][col] == num) {
            return false;
        }
    }
    int start_row = row - row % static_cast<int>(sqrt(size));
    int start_col = col - col % static_cast<int>(sqrt(size));
    for (int i = 0; i < static_cast<int>(sqrt(size)); i++) {
        for (int j = 0; j < static_cast<int>(sqrt(size)); j++) {
            if (board[i + start_row][j + start_col] == num) {
                return false;
            }
        }
    }
    return true;
}

bool solve_sudoku(std::vector<std::vector<int>>& board, int size) {
    int row = -1, col = -1;
    bool isEmpty = true;
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            if (board[i][j] == 0) {
                row = i;
                col = j;
                isEmpty = false;
                break;
            }
        }
        if (!isEmpty) {
            break;
        }
    }
    if (isEmpty) {
        return true;
    }
    for (int num = 1; num <= size; num++) {
        if (is_valid(board, row, col, num, size)) {
            board[row][col] = num;
            if (solve_sudoku(board, size)) {
                return true;
            }
            board[row][col] = 0;
        }
    }
    return false;
}

void print_board(const std::vector<std::vector<int>>& board) {
    for (const auto& row : board) {
        for (int num : row) {
            std::cout << num << " ";
        }
        std::cout << std::endl;
    }
}

int main() {
    int size;
    std::cin >> size;
    std::vector<std::vector<int>> board(size, std::vector<int>(size));
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            std::cin >> board[i][j];
        }
    }
    if (solve_sudoku(board, size)) {
        print_board(board);
    }
    else {
        std::cout << "wujie" << std::endl;
    }
    return 0;
}

File: test-2025.9.2/编程/数独/数独.py
def is_valid(board, row, col, num, size):
    for i in range(size):
        if board[row][i] == num or board[i][col] == num:
            return False
    start_row, start_col = row - row % int(size**0.5), col - col % int(size**0.5)
    for i in range(int(size**0.5)):
        for j in range(int(size**0.5)):
            if board[i + start_row][j + start_col] == num:
                return False
    return True

def solve_sudoku(board, size):
    empty = None
    for i in range(size):
        for j in range(size):
            if board[i][j] == 0:
                empty = (i, j)
                break
        if empty:
            break
    if not empty:
        return True
    row, col = empty
    for num in range(1, size + 1):
        if is_valid(board, row, col, num, size):
            board[row][col] = num
            if solve_sudoku(board, size):
                return True
            board[row][col] = 0
    return False

def print_board(board):
    for row in board:
        print(' '.join(map(str, row)))

def main():
    size = int(input())
    board = []
    for _ in range(size):
        board.append(list(map(int, input().split())))
    if solve_sudoku(board, size):
        print_board(board)
    else:
        print("No solution exists")

if __name__ == "__main__":
    main()